packages/backend-storage/src/storage_access_arbiter.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  ResourceAccessAcceptor,
+  SsmEnvironmentEntry,
+} from '@aws-amplify/plugin-types';
+import { StorageAction, StoragePath } from './types.js';
+import { StorageAccessPolicyFactory } from './storage_access_policy_factory.js';
+
+// some types internal to this file to improve readability
+type AcceptorToken = string;
+type SetDenyByDefault = (storagePath: StoragePath) => void;
+
+/**
+ * Internal collaborating class for building up mappings of the actions and s3 prefix resource scopes that should be granted to each ResourceAccessAcceptors.
+ * This class maintains two different data structures on top of the same underlying values.
+ * The first is a "forward map" of the allow/deny rules for each access acceptor
+ * The second if a "reverse map" of a "DenyByDefault" callback list for each s3 prefix that has been seen.
+ * This callback is used during a final pass over all of the rules to update the forward map with deny rules on any paths that do not have explicit allow rules
+ */
+export class StorageAccessArbiter {
+  /**
+   * Maintains a mapping from a resource access acceptor to all of the access grants it has been configured with
+   * Each entry of this map is fed into the policy generator to create a single policy for each acceptor
+   */
+  private acceptorAccessMap = new Map<
+    AcceptorToken,
+    {
+      acceptor: ResourceAccessAcceptor;
+      accessMap: Map<
+        StorageAction,
+        { allow: Set<StoragePath>; deny: Set<StoragePath> }
+      >;
+    }
+  >();
+
+  /**
+   * Maintains pointers to the "deny" StoragePath Set for each access entry in the map above
+   * This map is used during a final pass over all the StoragePaths to deny access on any paths where explicit allow rules were not specified
+   */
+  private prefixDenyMap = new Map<StoragePath, SetDenyByDefault[]>();
+
+  /**
+   * Initialize with a policyFactory.
+   * The policyFactory is the component that actually knows how to construct IAM policies.
+   * This class only manages the defineStorage access pattern abstraction
+   */
+  constructor(private readonly policyFactory: StorageAccessPolicyFactory) {}
+
+  /**
+   * Update the translator with an access definition.
+   * This definition defines a set of actions on a single s3 prefix that should be attached to a given ResourceAccessAcceptor
+   */
+  addAccessDefinition = (
```

Comment:
this method is where we setup the data structures for figuring out prefix deny rules when creating policies